File: server/src/main/java/com/uten/imp/features/production/plan/ProductionPlan.java
package com.uten.imp.features.production.plan;

import com.uten.imp.common.domain.SoftDeletableEntity;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.Table;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

import java.time.LocalDate;
import java.util.UUID;

/**
 * 生产计划单头（生产管理 · 源 F_Plan，7,235 行）。
 *
 * <p>审核状态机（design §4.1）：status 0 草稿 / 1 已审 / -1 红冲。审核（0→1）：
 * <ul>
 *   <li>重算 {@link #isClosed}（CheckFulfill4 派生，所有明细 qty - iqty ≤ 0）</li>
 *   <li>【本期后置】回写 sales_order_items 的 PQTY/LQTY/PlanNo（销售模块上线后）</li>
 *   <li>【本期后置】设 plan_items.step_legacy_id 首工序（车间模块上线后）</li>
 *   <li>【本期后置】填充 F_ProductingItem 按日产能（排产模块上线后）</li>
 * </ul>
 *
 * <p><b>不调</b> {@code StockService}（计划不动库存）<b>不调</b> {@code ArApLedgerService}（计划不立帐）。
 *
 * <p>车间字段：老库 WorkShop varchar(250) 装数字/名字（样本 "37"/"38"），新库 {@link #departmentId}
 * UUID 能对齐才填，{@link #workshopName} 原样文本留底（待 workshop_legacy_map 回填）。
 */
@Getter
@Setter
@NoArgsConstructor
@Entity
@Table(name = "production_plans")
public class ProductionPlan extends SoftDeletableEntity {

    @Column(name = "legacy_id", unique = true)
    private Integer legacyId;

    @Column(name = "bill_no", nullable = false)
    private String billNo;

    @Column(name = "bill_date", nullable = false)
    private LocalDate billDate;

    /** FStyle 生产类型（老库样本为空）。 */
    @Column(name = "f_style")
    private String fStyle;

    /** DDate 交货日期。 */
    @Column(name = "delivery_date")
    private LocalDate deliveryDate;

    /** WorkShop → departments（能对齐才填；老库 varchar 装名字/数字）。 */
    @Column(name = "department_id")
    private UUID departmentId;

    /** WorkShop varchar(250) 原样留底（车间编号/名字字符串）。 */
    @Column(name = "workshop_name")
    private String workshopName;

    /** WorkerID varchar(250) 多值名字，原样文本。 */
    @Column(name = "worker_name")
    private String workerName;

    /** Seller varchar(250) 跟单员，原样文本。 */
    @Column(name = "seller_name")
    private String sellerName;

    /** MakeID → employees（迁移留空，B_Worker 与 employees 无 legacy_id 对齐）。 */
    @Column(name = "maker_id")
    private UUID makerId;

    @Column(name = "approver_id")
    private UUID approverId;

    /** MakeID 老 ID 留底（后续 worker_legacy_map 回填）。 */
    @Column(name = "maker_legacy_id")
    private Integer makerLegacyId;

    @Column(name = "approver_legacy_id")
    private Integer approverLegacyId;

    /** Remark（text 大字段）。 */
    private String remark;

    /** 0 草稿 / 1 已审 / -1 红冲。 */
    @Column(name = "status", nullable = false)
    private Short status = 0;

    /** Fulfill 派生（CheckFulfill4 → Service 派生：所有明细 qty - iqty ≤ 0）。 */
    @Column(name = "is_closed", nullable = false)
    private boolean closed = false;

    /** Stop（手工中止）。 */
    @Column(name = "is_stopped", nullable = false)
    private boolean stopped = false;

    /** Cancel（取消）。 */
    @Column(name = "is_canceled", nullable = false)
    private boolean canceled = false;

    /** 软关联占位（销售订单号等）。 */
    @Column(name = "source_doc_no")
    private String sourceDocNo;
}

File: lib/features/production/models/production_plan.dart
// 生产计划单 model（生产管理 / production）。
//
// 对应后端 server/src/main/java/com/uten/imp/features/production/plan/：
//   ProductionPlan（头）+ ProductionPlanItem（明细，41 字段含 12 数量族）。
// DTO 形状：PlanListItem（列表行）/ PlanDetail（详情含 items）/ PlanItemDto（明细行）。
//
// 状态机：0=草稿 / 1=已审 / -1=红冲（与采购/日报一致；记忆 purchase-module-progress）。
// JSON 注意：Jackson 把 boolean isClosed/isStopped/isCanceled 序列化为 closed/stopped/canceled
//   （去掉 is 前缀）；日期为 yyyy-MM-dd 字符串；数量/金额 NUMERIC(18,4) 按 num? 容错。
// 容错助手 _asInt/_asDouble 同时处理 int/double/String（记忆 flutter-int-cast-fromjson）。
import 'package:flutter/material.dart';

/// 生产单据状态常量（plan/daily 共用；与后端 status SMALLINT 0/1/-1 对齐）。
const int kProductionStatusDraft = 0;
const int kProductionStatusApproved = 1;
const int kProductionStatusReversed = -1;

String productionStatusLabel(int? code) {
  switch (code) {
    case kProductionStatusDraft:
      return '草稿';
    case kProductionStatusApproved:
      return '已审';
    case kProductionStatusReversed:
      return '红冲';
    default:
      return '—';
  }
}

/// 截取日期字符串为 yyyy-MM-dd；null/空/不足 10 位则原样返回（避免 substring 越界：
/// 后端日期可能为 null 或短串，`''.substring(0,10)` 会抛 RangeError 整页渲染失败）。
String productionDateOnly(String? s) {
  if (s == null) return '';
  return s.length >= 10 ? s.substring(0, 10) : s;
}

Color productionStatusColor(int? code, ThemeData theme) {
  switch (code) {
    case kProductionStatusApproved:
      return Colors.green;
    case kProductionStatusReversed:
      return theme.colorScheme.error;
    default:
      return theme.colorScheme.onSurfaceVariant;
  }
}

int? _asInt(dynamic v) {
  if (v == null) return null;
  if (v is int) return v;
  if (v is num) return v.toInt();
  if (v is String) return int.tryParse(v);
  return null;
}

double? _asDouble(dynamic v) {
  if (v == null) return null;
  if (v is double) return v;
  if (v is num) return v.toDouble();
  if (v is String) return double.tryParse(v);
  return null;
}

/// 生产计划单列表行（GET /production/plans → PlanListItem）。
class ProductionPlanListItem {
  const ProductionPlanListItem({
    required this.id,
    this.billNo,
    this.billDate,
    this.deliveryDate,
    this.departmentId,
    this.workshopName,
    this.workerName,
    this.sellerName,
    this.status,
    this.closed = false,
    this.stopped = false,
    this.canceled = false,
    this.legacyId,
  });

  final String id;
  final String? billNo;
  final String? billDate;
  final String? deliveryDate;
  final String? departmentId;
  final String? workshopName;
  final String? workerName;
  final String? sellerName;
  final int? status;
  final bool closed;
  final bool stopped;
  final bool canceled;
  final int? legacyId;

  factory ProductionPlanListItem.fromJson(Map<String, dynamic> json) =>
      ProductionPlanListItem(
        id: json['id'] as String,
        billNo: json['billNo'] as String?,
        billDate: json['billDate'] as String?,
        deliveryDate: json['deliveryDate'] as String?,
        departmentId: json['departmentId'] as String?,
        workshopName: json['workshopName'] as String?,
        workerName: json['workerName'] as String?,
        sellerName: json['sellerName'] as String?,
        status: _asInt(json['status']),
        closed: (json['closed'] as bool?) ?? false,
        stopped: (json['stopped'] as bool?) ?? false,
        canceled: (json['canceled'] as bool?) ?? false,
        legacyId: _asInt(json['legacyId']),
      );
}

/// 生产计划明细行（PlanItemDto，含 12 数量族；BOM 展开挂 billItemId→本 id）。
class ProductionPlanItem {
  const ProductionPlanItem({
    required this.id,
    this.lineNo,
    this.productNo,
    this.goodsId,
    this.colorId,
    this.mgoodsId,
    this.unitId,
    this.unitRate,
    this.salesOrderItemId,
    this.salesOrderNo,
    this.clientName,
    this.clientNo,
    this.oqty,
    this.qty,
    this.lqty,
    this.iqty,
    this.fqty,
    this.rqty,
    this.bqty,
    this.tqty,
    this.paqty,
    this.isrqty,
    this.cpqty,
    this.poqty,
    this.piqty,
    this.orderDate,
    this.outboundDate,
    this.planBeginDate,
    this.planEndDate,
    this.finishedWeight,
    this.inboundWeight,
    this.lstatus,
    this.cstatus,
    this.stepLegacyId,
    this.veilLegacyId,
    this.assTeamLegacyId,
    this.fittings,
    this.requestNote,
    this.customerModel,
    this.discount,
    this.labelNo,
    this.planAppNo,
    this.sourceDocNo,
    this.remark,
  });

  final String id;
  final int? lineNo;
  final String? productNo;
  final String? goodsId;
  final String? colorId;
  final String? mgoodsId;
  final String? unitId;
  final double? unitRate;
  final String? salesOrderItemId;
  final String? salesOrderNo;
  final String? clientName;
  final String? clientNo;

  /// 数量族（12 个，NUMERIC(18,4)；触发器游标回写累计量，本期不重算）。
  final double? oqty; // 销售订货量
  final double? qty; // 本单排产数量
  final double? lqty; // BOM 展开锁定用量
  final double? iqty; // 完工/进仓数量（仓库回写）
  final double? fqty; // 完工数量（工序回写）
  final double? rqty; // 入库数量
  final double? bqty; // 在产数量
  final double? tqty; // 开工数量
  final double? paqty; // 已排产量
  final double? isrqty; // 已入库量
  final double? cpqty; // 应排数量
  final double? poqty; // 已订货（采购回写）
  final double? piqty; // 已收货（采购回写）

  final String? orderDate;
  final String? outboundDate;
  final String? planBeginDate;
  final String? planEndDate;
  final double? finishedWeight;
  final double? inboundWeight;
  final int? lstatus;
  final int? cstatus;
  final int? stepLegacyId;
  final int? veilLegacyId;
  final int? assTeamLegacyId;
  final String? fittings;
  final String? requestNote;
  final String? customerModel;
  final double? discount;
  final String? labelNo;
  final String? planAppNo;
  final String? sourceDocNo;
  final String? remark;

  factory ProductionPlanItem.fromJson(Map<String, dynamic> json) =>
      ProductionPlanItem(
        id: json['id'] as String,
        lineNo: _asInt(json['lineNo']),
        productNo: json['productNo'] as String?,
        goodsId: json['goodsId'] as String?,
        colorId: json['colorId'] as String?,
        mgoodsId: json['mgoodsId'] as String?,
        unitId: json['unitId'] as String?,
        unitRate: _asDouble(json['unitRate']),
        salesOrderItemId: json['salesOrderItemId'] as String?,
        salesOrderNo: json['salesOrderNo'] as String?,
        clientName: json['clientName'] as String?,
        clientNo: json['clientNo'] as String?,
        oqty: _asDouble(json['oqty']),
        qty: _asDouble(json['qty']),
        lqty: _asDouble(json['lqty']),
        iqty: _asDouble(json['iqty']),
        fqty: _asDouble(json['fqty']),
        rqty: _asDouble(json['rqty']),
        bqty: _asDouble(json['bqty']),
        tqty: _asDouble(json['tqty']),
        paqty: _asDouble(json['paqty']),
        isrqty: _asDouble(json['isrqty']),
        cpqty: _asDouble(json['cpqty']),
        poqty: _asDouble(json['poqty']),
        piqty: _asDouble(json['piqty']),
        orderDate: json['orderDate'] as String?,
        outboundDate: json['outboundDate'] as String?,
        planBeginDate: json['planBeginDate'] as String?,
        planEndDate: json['planEndDate'] as String?,
        finishedWeight: _asDouble(json['finishedWeight']),
        inboundWeight: _asDouble(json['inboundWeight']),
        lstatus: _asInt(json['lstatus']),
        cstatus: _asInt(json['cstatus']),
        stepLegacyId: _asInt(json['stepLegacyId']),
        veilLegacyId: _asInt(json['veilLegacyId']),
        assTeamLegacyId: _asInt(json['assTeamLegacyId']),
        fittings: json['fittings'] as String?,
        requestNote: json['requestNote'] as String?,
        customerModel: json['customerModel'] as String?,
        discount: _asDouble(json['discount']),
        labelNo: json['labelNo'] as String?,
        planAppNo: json['planAppNo'] as String?,
        sourceDocNo: json['sourceDocNo'] as String?,
        remark: json['remark'] as String?,
      );
}

/// 生产计划单详情（GET /production/plans/{id} → PlanDetail）。
class ProductionPlanDetail {
  const ProductionPlanDetail({
    required this.id,
    this.legacyId,
    this.billNo,
    this.billDate,
    this.fStyle,
    this.deliveryDate,
    this.departmentId,
    this.workshopName,
    this.workerName,
    this.sellerName,
    this.makerId,
    this.approverId,
    this.makerLegacyId,
    this.approverLegacyId,
    this.remark,
    this.status,
    this.closed = false,
    this.stopped = false,
    this.canceled = false,
    this.sourceDocNo,
    this.items = const [],
  });

  final String id;
  final int? legacyId;
  final String? billNo;
  final String? billDate;
  final String? fStyle;
  final String? deliveryDate;
  final String? departmentId;
  final String? workshopName;
  final String? workerName;
  final String? sellerName;
  final String? makerId;
  final String? approverId;
  final int? makerLegacyId;
  final int? approverLegacyId;
  final String? remark;
  final int? status;
  final bool closed;
  final bool stopped;
  final bool canceled;
  final String? sourceDocNo;
  final List<ProductionPlanItem> items;

  factory ProductionPlanDetail.fromJson(Map<String, dynamic> json) =>
      ProductionPlanDetail(
        id: json['id'] as String,
        legacyId: _asInt(json['legacyId']),
        billNo: json['billNo'] as String?,
        billDate: json['billDate'] as String?,
        fStyle: json['fStyle'] as String?,
        deliveryDate: json['deliveryDate'] as String?,
        departmentId: json['departmentId'] as String?,
        workshopName: json['workshopName'] as String?,
        workerName: json['workerName'] as String?,
        sellerName: json['sellerName'] as String?,
        makerId: json['makerId'] as String?,
        approverId: json['approverId'] as String?,
        makerLegacyId: _asInt(json['makerLegacyId']),
        approverLegacyId: _asInt(json['approverLegacyId']),
        remark: json['remark'] as String?,
        status: _asInt(json['status']),
        closed: (json['closed'] as bool?) ?? false,
        stopped: (json['stopped'] as bool?) ?? false,
        canceled: (json['canceled'] as bool?) ?? false,
        sourceDocNo: json['sourceDocNo'] as String?,
        items: (json['items'] as List?)
                ?.map((e) =>
                    ProductionPlanItem.fromJson(e as Map<String, dynamic>))
                .toList() ??
            const [],
      );
}
